### ISSUE TITLE
Schema Markup

### ISSUE THREAD

**Issue by rwcsrichardspencer** [aemsites/stericycle-shared#363]
Attached is a list of our current schema set up on the website. Ideally we'd like everything to be "Auto" where the schema JSON-LD is built using in-page or metadata information/values. Or suggest key/value pairs in the metadata to be able to build out the JSON-LD.

[Website schema table.xlsx](https://github.com/user-attachments/files/17436305/Website.schema.table.xlsx)

**Comment by kronnox** [comment]
This can for sure be added automatically to all places where the JSON-LD is needed, as long as some kind of pattern is present and all the necessary information is present on the page / within the metadata. From what I've seen on other projects this is usually either done on a page or a block level. On a page level this could be triggered through the media-type (e.g. for Blogs) and on a block level just every time the block is used (e.g. FAQ).

@rwcsrichardspencer Is there currently a list of all the places where JSON-LD is added so far? For some of them it's obvious to me from the description (Video, FAQ, Breadcrumb), for others I'm not sure where this is currently in use (Organization, SiteNavigationElement, WebPage, ...). It would help me out tremendously if a had a list of examples or even better a mapping of page-type / block to the corresponding JSON-LD and where the data in there should be derived from.

**Comment by kronnox** [comment]
Example of how to add JSON-LD to a page:
```js
function setJsonLdProduct(product) {
  const {
    name, inStock, description, sku, price, priceRange, productImages, attributes,
  } = product;
  const amount = priceRange?.minimum?.final?.amount || price?.final?.amount;
  const brand = attributes.find((attr) => attr.name === 'brand');
  setJsonLd({
    '@context': 'http://schema.org',
    '@type': 'Product',
    name,
    description,
    image: productImages[0]?.url,
    offers: [{
      '@type': 'http://schema.org/Offer',
      price: amount?.value,
      priceCurrency: amount?.currency,
      availability: inStock ? 'http://schema.org/InStock' : 'http://schema.org/OutOfStock',
    }],
    productID: sku,
    brand: {
      '@type': 'Brand',
      name: brand?.value,
    },
    url: `https://www.example.com/products/${product.url_key}/${sku}`,
    sku,
    '@id': `https://www.example.com/products/${product.url_key}/${sku}`,
  }, 'product');
}
```
```js
export function setJsonLd(data, name) {
  const existingScript = document.head.querySelector(`script[data-name="${name}"]`);
  if (existingScript) {
    existingScript.innerHTML = JSON.stringify(data);
    return;
  }
  const script = document.createElement('script');
  script.type = 'application/ld+json';
  script.innerHTML = JSON.stringify(data);
  script.dataset.name = name;
  document.head.appendChild(script);
}
```

**Comment by rwcsrichardspencer** [comment]
I provided a sheet that shows what we use and what's automated vs manual (copy/paste JSON+LD). I think a sheet like the metadata or query-index might be the way to go so it's all managed in 1 place and can hold patterns by just using the column titles and still have the ability to add our own JSON in the metadata table of any given page. Any thoughts on this approach?

**Comment by kronnox** [comment]
From a technical point-of-view it should be fine to have this managed through a sheet. I think for the manual ones it's just a design preference whether you want all of them in one sheet or to have the schema in the document itself; no preference from my side. The automatically deployed schemas would be handled through JavaScript.

**Comment by rwcsrichardspencer** [comment]
Found VideoObject:
<script class="w-json-ld" type="application/ld+json" id="w-json-ldwistia_40">{"@context":"
[http://schema.org/","@id":"https://fast.wistia.net/embed/iframe/t2cnjqnm6s","@type":"VideoObject","duration":"PT58S","name":"Shred-it](http://schema.org/%22,%22@id%22:%22https://fast.wistia.net/embed/iframe/t2cnjqnm6s%22,%22@type%22:%22VideoObject%22,%22duration%22:%22PT58S%22,%22name%22:%22Shred-it)
Generic Video","thumbnailUrl":"
[https://embed-ssl.wistia.com/deliveries/b11738bca47ead75483883098d6b9f51.jpg?image_crop_resized=640x358","embedUrl":"https://fast.wistia.net/embed/iframe/t2cnjqnm6s","uploadDate":"2021-08-05T19:28:31.000Z","description":"Title:](https://embed-ssl.wistia.com/deliveries/b11738bca47ead75483883098d6b9f51.jpg?image_crop_resized=640x358%22,%22embedUrl%22:%22https://fast.wistia.net/embed/iframe/t2cnjqnm6s%22,%22uploadDate%22:%222021-08-05T19:28:31.000Z%22,%22description%22:%22Title:)
Shred-it’s Information Security & Document Destruction\nDescription: With Shred-it’s 30 years of experience securing information, you get the security and convenience for your business needs for your paper shredding and hard drive destruction. Our paper and hard drive shredding services are NAID AAA-certified to keep your information safe. Protect your confidential information, protect your firm. Get Shred-it.","contentUrl":"
[https://embed-ssl.wistia.com/deliveries/a65b2b35d606c4e3ad8d11634be956451010829d.m3u8","transcript":"(Music](https://embed-ssl.wistia.com/deliveries/a65b2b35d606c4e3ad8d11634be956451010829d.m3u8%22,%22transcript%22:%22(Music)
playing)","potentialAction":{"@type":"SeekToAction","target":"
[https://www.shredit.co.uk/en-gb/industries/healthcare?wtime={seek_to_second_number}","startOffset-input":"required](https://www.shredit.co.uk/en-gb/industries/healthcare?wtime=%7Bseek_to_second_number%7D%22,%22startOffset-input%22:%22required)
name=seek_to_second_number"}}</script><meta http-equiv="origin-trial" content="A/kargTFyk8MR5ueravczef/wIlTkbVk1qXQesp39nV+xNECPdLBVeYffxrM8TmZT6RArWGQVCJ0LRivD7glcAUAAACQeyJvcmlnaW4iOiJodHRwczovL2dvb2dsZS5jb206NDQzIiwiZmVhdHVyZSI6IkRpc2FibGVUaGlyZFBhcnR5U3RvcmFnZVBhcnRpdGlvbmluZzIiLCJleHBpcnkiOjE3NDIzNDIzOTksImlzU3ViZG9tYWluIjp0cnVlLCJpc1RoaXJkUGFydHkiOnRydWV9" />

**Comment by rwcsrichardspencer** [comment]
@kronnox @MarquiseRosier it's one thing to mark as done, but please provide some insight into what was done and how it works so we can take a look.

**Comment by rwcsrichardspencer** [comment]
@kronnox looks like schema added via page metadata is escaping the quote characters with backslash, thus rendering the schema unintelligible by crawlers.
![quoteescape](https://github.com/user-attachments/assets/772035a9-d66d-4457-a7f7-b505055baf69)
Can we resolve this quickly?

**Comment by kronnox** [comment]
@rwcsrichardspencer You're absolutely right. I proposed a fix in #509, but I need to reevaluate if this poses a security issue.
